file/cache.lisp

```diff
@@ -36,55 +44,93 @@
-             :end1 (+ (file-position* stream) (- end start))
-             :start2 start
-             :end2 end2)
-    (incf (file-position* stream) (- end2 start))))
+  (assert (input-stream-p stream))
+  (if (>= (file-%position stream) (file-%length stream))
+      :eof
+      (multiple-value-bind (block-n block-offset)
+          (floor (file-%position stream) (file-%block-size stream))
+        (when (/= block-n (file-%block-n stream))
+          (finish-output stream)
+          (setf (file-%block-n stream) block-n
+                (file-%buffer stream) (read-file-block stream block-n)))
+        (prog1 (aref (file-%buffer stream) block-offset)
+          (incf (file-%position stream))))))
+
+;; TODO
+;; (defmethod mezzano.gray:stream-write-sequence ((stream file-cache-stream) sequence &optional (start 0) end)
+;;   (assert (output-stream-p stream))
+;;   (let* ((end (or end (file-%length stream)))
+;;          (end2 (min end (length sequence))))
+;;     (when (> end2 (file-%length stream))
+;;       (setf (file-%buffer stream) (adjust-array (file-%buffer stream) end2 :initial-element 0)))
+;;     (replace (file-%buffer stream) sequence
+;;              :start1 (file-%position stream)
+;;              :end1 (+ (file-%position stream) (- end start))
+;;              :start2 start
+;;              :end2 end2)
+;;     (incf (file-%position stream) (- end2 start))))
 
 (defmethod mezzano.gray:stream-read-sequence ((stream file-cache-stream) sequence &optional (start 0) end)
-  (assert (member (direction stream) '(:input :io)))
-  (let* ((end (or end (length sequence)))
-         (end2 (min end (file-length* stream))))
-    (replace sequence (buffer stream) :start1 start :end1 end :start2 0 :end2 end2)
-    end2))
+  (assert (input-stream-p stream))
+  (let* ((buffer-length (file-%block-size stream))
+         (end1 (or end (length sequence)))
+         (end2 (+ (file-%position stream) (min end1 (file-%length stream)))))
+    (multiple-value-bind (first-block-n start-offset)
+        (floor (file-%position stream) buffer-length)
+      (multiple-value-bind (last-block-n end-offset)
+          (floor (- end2 start) buffer-length)
+        (unless (= (file-%block-n stream) first-block-n last-block-n)
+          (finish-output stream))
+        (let ((block-n first-block-n))
+          (do ((file-block (if (= first-block-n (file-%block-n stream))
+                               (file-%buffer stream)
+                               (read-file-block stream block-n))
+                           (read-file-block stream block-n))
+               (start2 start-offset 0)
+               (start1 start (+ start1 (- buffer-length start2))))
+              ((= last-block-n block-n)
+               (unless (zerop end-offset)
+                 (replace sequence file-block :start1 start1 :end1 end1 :start2 start2 :end2 end-offset))
+               (setf (file-%buffer stream) file-block
+                     (file-%block-n stream) block-n)
+               end2)
```

Comment:
And this would be `(+ (- end2 (file-%position stream)) start)` to get the final position in the sequence.
file-%position also needs to be updated after everything has been read.